src/hypothesis/strategies.py

```diff
@@ -1005,39 +1005,91 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
     """Returns a strategy which generates Fractions.
 
     If min_value is not None then all generated values are no less than
-    min_value.
+    min_value.  If max_value is not None then all generated values are no
+    greater than max_value.  min_value and max_value may be anything accepted
+    by the `python:`~fractions.Fraction` constructor.
 
-    If max_value is not None then all generated values are no greater than
-    max_value.
-
-    If max_denominator is not None then the absolute value of the denominator
-    of any generated values is no greater than max_denominator. Note that
-    max_denominator must be at least 1.
+    If max_denominator is not None then the denominator of any generated
+    values is no greater than max_denominator. Note that max_denominator must
+    be None or a positive integer.
 
     """
+    if min_value is not None:
+        min_value = Fraction(min_value)
+    if max_value is not None:
+        max_value = Fraction(max_value)
+
     check_valid_bound(min_value, 'min_value')
     check_valid_bound(max_value, 'max_value')
     check_valid_interval(min_value, max_value, 'min_value', 'max_value')
-
     check_valid_integer(max_denominator)
-    if max_denominator is not None and max_denominator < 1:
-        raise InvalidArgument(
-            u'Invalid denominator bound %s' % max_denominator
-        )
 
-    denominator_strategy = integers(min_value=1, max_value=max_denominator)
+    if max_denominator is not None:
+        if max_denominator < 1:
+            raise InvalidArgument(
+                'max_denominator=%r must be >= 1' % max_denominator)
+
+        def fraction_bounds(value):
+            """Find the best lower and upper approximation for value."""
+            # Adapted from CPython's Fraction.limit_denominator here:
+            # https://github.com/python/cpython/blob/3.6/Lib/fractions.py#L219
+            if value is None or value.denominator <= max_denominator:
+                return value, value
+            p0, q0, p1, q1 = 0, 1, 1, 0
+            n, d = value.numerator, value.denominator
+            while True:
+                a = n // d
+                q2 = q0 + a * q1
+                if q2 > max_denominator:
+                    break
+                p0, q0, p1, q1 = p1, q1, p0 + a * p1, q2
+                n, d = d, n - a * d
+            k = (max_denominator - q0) // q1
+            low, high = Fraction(p1, q1), Fraction(p0 + k * p1, q0 + k * q1)
+            assert low < value < high
+            return low, high
+
+        # Take the high approximation for min_value and low for max_value
+        bounds = (max_denominator, min_value, max_value)
+        _, min_value = fraction_bounds(min_value)
+        max_value, _ = fraction_bounds(max_value)
+
+        if None not in (min_value, max_value) and min_value > max_value:
+            raise InvalidArgument(
+                'There are no rational numbers with a denominator <= %r '
```

Comment:
Possibly better to use "fractions" rather than "rational numbers" to be consistent with the API.